README.md
# Mojo-Reactor-UV
Mojo::Reactor::UV - UV backend for Mojo::Reactor


Changes
{{$NEXT}}

1.901     2021-06-16 23:05:54 EDT
  - Stable release requiring UV 1.909

1.900     2021-06-16 01:23:42 EDT (TRIAL RELEASE)
  - Changes to support UV 2.0
  - Update IRC metadata to libera.chat

1.002     2020-04-19 21:40:34 EDT
  - Support changing invocation time of active timers in again method, to
    support keep_alive_timeout added in Mojolicious 8.37

1.001     2018-04-21 20:19:51 EDT
  - Stable release using UV 1.0
  - Use UV loops instead of Poll when multiple reactors are constructed

1.000     2018-01-26 04:02:07 EST (TRIAL RELEASE)
  - Switch installer to Module::Build::Tiny
  - Changes to support UV 1.0

0.001     2015-06-05 18:10:25 EDT
  - First release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.036.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "UV backend for Mojo::Reactor",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Mojo-Reactor-UV",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Mojo::Reactor::UV",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Mojolicious" => "6.0",
    "Scalar::Util" => 0,
    "UV" => "0.23"
  },
  "TEST_REQUIRES" => {
    "IO::Socket::INET" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "IO::Socket::INET" => 0,
  "Mojolicious" => "6.0",
  "Scalar::Util" => 0,
  "Test::More" => "0.88",
  "UV" => "0.23"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/Reactor/UV.pm
package Mojo::Reactor::UV;
use Mojo::Base 'Mojo::Reactor::Poll';

$ENV{MOJO_REACTOR} ||= 'Mojo::Reactor::UV';

use Carp 'croak';
use Mojo::Util qw(md5_sum steady_time);
use Scalar::Util qw(blessed weaken);
use UV;
use UV::Poll;
use UV::Timer;
use UV::Loop;

use constant DEBUG => $ENV{MOJO_REACTOR_UV_DEBUG} || 0;

our $VERSION = '1.902';

my $UV;

# Use UV::Loop singleton for the first instance only
sub new {
	my $self = shift->SUPER::new;
	if ($UV++) {
		$self->{loop} = UV::Loop->new;
	} else {
		$self->{loop} = UV::Loop->default_loop;
		$self->{loop_singleton} = 1;
	}
	return $self;
}

sub DESTROY { undef $UV if shift->{loop_singleton} }

sub again {
	my ($self, $id, $after) = @_;
	croak 'Timer not active' unless my $timer = $self->{timers}{$id};
	my $w = $timer->{watcher};
	if (defined $after) {
		$after *= 1000; # Intervals in milliseconds
		# Timer will not repeat with (integer) interval of 0
		$after = 1 if $after < 1;
		eval { $w->repeat($after); 1 } or $self->_error($@);
	}
	eval { $w->again; 1 } or $self->_error($@);
}

sub io {
	my ($self, $handle, $cb) = @_;
	my $fd = fileno($handle) // croak 'Handle is closed';
	# Must use existing watcher if present
	$self->{io}{$fd}{cb} = $cb;
	warn "-- Set IO watcher for $fd\n" if DEBUG;
	return $self->watch($handle, 1, 1);
}

sub one_tick {
	my $self = shift;
	# Just one tick
	local $self->{running} = 1 unless $self->{running};
	$self->{loop}->run(UV::Loop::UV_RUN_ONCE) or $self->stop;
}

sub recurring { shift->_timer(1, @_) }

sub remove {
	my ($self, $remove) = @_;
	return unless defined $remove;
	if (ref $remove) {
		my $fd = fileno($remove) // croak 'Handle is closed';
		if (exists $self->{io}{$fd}) {
			warn "-- Removed IO watcher for $fd\n" if DEBUG;
			my $w = delete $self->{io}{$fd}{watcher};
			$w->close if $w;
		}
		return !!delete $self->{io}{$fd};
	} else {
		if (exists $self->{timers}{$remove}) {
			warn "-- Removed timer $remove\n" if DEBUG;
			my $w = delete $self->{timers}{$remove}{watcher};
			$w->close if $w;
		}
		return !!delete $self->{timers}{$remove};
	}
}

sub reset {
	my $self = shift;
	$_->close for map { $_->{watcher} ? ($_->{watcher}) : () }
		values %{$self->{io}}, values %{$self->{timers}};
	$self->SUPER::reset;
}

sub timer { shift->_timer(0, @_) }

sub watch {
	my ($self, $handle, $read, $write) = @_;
	
	my $fd = fileno $handle;
	croak 'I/O watcher not active' unless my $io = $self->{io}{$fd};
	
	my $mode = 0;
	$mode |= UV::Poll::UV_READABLE if $read;
	$mode |= UV::Poll::UV_WRITABLE if $write;
	
	my $w;
	unless ($w = $io->{watcher}) { $w = $io->{watcher} = UV::Poll->new(loop => $self->{loop}, fd => $fd); }
	
	if ($mode == 0) { eval { $w->stop; 1 } or $self->_error($@); }
	else {
		weaken $self;
		my $cb = sub {
			my ($w, $status, $events) = @_;
			return $self->_error($status) if $status < 0;
			$self->_try('I/O watcher', $self->{io}{$fd}{cb}, 0)
				if UV::Poll::UV_READABLE & $events;
			$self->_try('I/O watcher', $self->{io}{$fd}{cb}, 1)
				if UV::Poll::UV_WRITABLE & $events && $self->{io}{$fd};
		};
		eval { $w->start($mode, $cb); 1 } or $self->_error($@);
	}
	
	return $self;
}

sub _error {
	my ($self, $err) = @_;
	if (blessed $err and $err->isa('UV::Exception')) {
		$self->emit(error => sprintf 'UV error: %s', $err->message);
	} elsif (!ref $err and $err < 0) {
		$self->emit(error => sprintf 'UV error: %s', UV::strerror($err));
	}
	return $err;
}

sub _id {
	my $self = shift;
	my $id;
	do { $id = md5_sum 't' . steady_time . rand } while $self->{timers}{$id};
	return $id;
}

sub _timer {
	my ($self, $recurring, $after, $cb) = @_;
	$after *= 1000; # Intervals in milliseconds
	my $recur_after = $after;
	# Timer will not repeat with (integer) interval of 0
	$recur_after = 1 if $recurring and $after < 1;
	
	my $id = $self->_id;
	weaken $self;
	my $wrapper = sub {
		$self->remove($id) unless $recurring;
		$self->_try('Timer', $cb);
	};
	my $w = $self->{timers}{$id}{watcher} = UV::Timer->new(loop => $self->{loop});
	eval { $w->start($after, $recur_after, $wrapper); 1 } or $self->_error($@);
	
	if (DEBUG) {
		my $is_recurring = $recurring ? ' (recurring)' : '';
		my $seconds = $after / 1000;
		warn "-- Set timer $id after $seconds seconds$is_recurring\n";
	}
		
	return $id;
}

sub _try {
	my ($self, $what, $cb) = (shift, shift, shift);
	eval { $self->$cb(@_); 1 } or $self->emit(error => "$what failed: $@");
}

1;

=encoding utf8

=head1 NAME

Mojo::Reactor::UV - UV backend for Mojo::Reactor

=head1 SYNOPSIS

  use Mojo::Reactor::UV;

  # Watch if handle becomes readable or writable
  my $reactor = Mojo::Reactor::UV->new;
  $reactor->io($first => sub {
    my ($reactor, $writable) = @_;
    say $writable ? 'First handle is writable' : 'First handle is readable';
  });

  # Change to watching only if handle becomes writable
  $reactor->watch($first, 0, 1);

  # Turn file descriptor into handle and watch if it becomes readable
  my $second = IO::Handle->new_from_fd($fd, 'r');
  $reactor->io($second => sub {
    my ($reactor, $writable) = @_;
    say $writable ? 'Second handle is writable' : 'Second handle is readable';
  })->watch($second, 1, 0);

  # Add a timer
  $reactor->timer(15 => sub {
    my $reactor = shift;
    $reactor->remove($first);
    $reactor->remove($second);
    say 'Timeout!';
  });

  # Start reactor if necessary
  $reactor->start unless $reactor->is_running;

  # Or in an application using Mojo::IOLoop
  use Mojo::Reactor::UV;
  use Mojo::IOLoop;

  # Or in a Mojolicious application
  $ MOJO_REACTOR=Mojo::Reactor::UV hypnotoad script/myapp

=head1 DESCRIPTION

L<Mojo::Reactor::UV> is an event reactor for L<Mojo::IOLoop> that uses
C<libuv>. The usage is exactly the same as other L<Mojo::Reactor>
implementations such as L<Mojo::Reactor::Poll>. L<Mojo::Reactor::UV> will be
used as the default backend for L<Mojo::IOLoop> if it is loaded before
L<Mojo::IOLoop> or any module using the loop. However, when invoking a
L<Mojolicious> application through L<morbo> or L<hypnotoad>, the reactor must
be set as the default by setting the C<MOJO_REACTOR> environment variable to
C<Mojo::Reactor::UV>.

=head1 EVENTS

L<Mojo::Reactor::UV> inherits all events from L<Mojo::Reactor::Poll>.

=head1 METHODS

L<Mojo::Reactor::UV> inherits all methods from L<Mojo::Reactor::Poll> and
implements the following new ones.

=head2 new

  my $reactor = Mojo::Reactor::UV->new;

Construct a new L<Mojo::Reactor::UV> object.

=head2 again

  $reactor->again($id);
  $reactor->again($id, 0.5);

Restart timer and optionally change the invocation time. Note that this method
requires an active timer.

=head2 io

  $reactor = $reactor->io($handle => sub {...});

Watch handle for I/O events, invoking the callback whenever handle becomes
readable or writable.

  # Callback will be invoked twice if handle becomes readable and writable
  $reactor->io($handle => sub {
    my ($reactor, $writable) = @_;
    say $writable ? 'Handle is writable' : 'Handle is readable';
  });

=head2 one_tick

  $reactor->one_tick;

Run reactor until an event occurs or no events are being watched anymore. Note
that this method can recurse back into the reactor, so you need to be careful.

  # Don't block longer than 0.5 seconds
  my $id = $reactor->timer(0.5 => sub {});
  $reactor->one_tick;
  $reactor->remove($id);

=head2 recurring

  my $id = $reactor->recurring(0.25 => sub {...});

Create a new recurring timer, invoking the callback repeatedly after a given
amount of time in seconds.

=head2 remove

  my $bool = $reactor->remove($handle);
  my $bool = $reactor->remove($id);

Remove handle or timer.

=head2 reset

  $reactor->reset;

Remove all handles and timers.

=head2 timer

  my $id = $reactor->timer(0.5 => sub {...});

Create a new timer, invoking the callback after a given amount of time in
seconds.

=head2 watch

  $reactor = $reactor->watch($handle, $readable, $writable);

Change I/O events to watch handle for with true and false values. Note that
this method requires an active I/O watcher.

  # Watch only for readable events
  $reactor->watch($handle, 1, 0);

  # Watch only for writable events
  $reactor->watch($handle, 0, 1);

  # Watch for readable and writable events
  $reactor->watch($handle, 1, 1);

  # Pause watching for events
  $reactor->watch($handle, 0, 0);

=head1 CAVEATS

When using L<Mojo::IOLoop> with L<UV>, the event loop must be controlled by
L<Mojo::IOLoop> or L<Mojo::Reactor::UV>, such as with the methods
L<Mojo::IOLoop/"start">, L<Mojo::IOLoop/"stop">, and L</"one_tick">. Starting
or stopping the event loop through L<UV> will not provide required
functionality to L<Mojo::IOLoop> applications.

Care should be taken that file descriptors are not closed while being watched
by the reactor. They can be safely closed after calling L</"watch"> with
C<readable> and C<writable> set to 0, or after removing the handle with
L</"remove"> or L</"reset">.

On windows, C<libuv> can only watch sockets, not regular filehandles.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book, C<dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright 2015, Dan Book.

This library is free software; you may redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Mojolicious>, L<Mojo::IOLoop>, L<UV>


Build.PL
# This Build.PL for Mojo-Reactor-UV was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for Mojo-Reactor-UV was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.010001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.017.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "UV backend for Mojo::Reactor",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "Mojo-Reactor-UV",
      "dist_version" => "1.900",
      "license" => "artistic_2",
      "module_name" => "Mojo::Reactor::UV",
      "recursive_test_files" => 1,
      "requires" => {
        "Carp" => 0,
        "Mojolicious" => "7.37",
        "Scalar::Util" => 0,
        "UV" => "1.900",
        "perl" => "5.010001"
      },
      "test_requires" => {
        "File::Spec" => 0,
        "IO::Socket::INET" => 0,
        "Module::Metadata" => 0,
        "Scalar::Util" => 0,
        "Test::More" => "0.88",
        "Test::Needs" => 0
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "IO::Socket::INET" => 0,
      "Module::Metadata" => 0,
      "Scalar::Util" => 0,
      "Test::More" => "0.88",
      "Test::Needs" => 0
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}
